pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -138,3 +123,71 @@ func TestReaderPool_ShouldCloseIdleLazyReaders(t *testing.T) {
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
 	require.Equal(t, float64(2), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
 }
+
+func TestReaderPool_PersistLazyLoadedBlock(t *testing.T) {
+	const idleTimeout = time.Second
+	ctx, tmpDir, bkt, blockID, metrics := prepareReaderPool(t)
+	defer func() { require.NoError(t, os.RemoveAll(tmpDir)) }()
+	defer func() { require.NoError(t, bkt.Close()) }()
+
+	lazyLoadedTracker := HeadersLazyLoadedTracker{
+		// Path stores where lazy loaded blocks will be tracked in per-tenant-files
+		Path: filepath.Join(tmpDir, "lazy-loaded.proto"),
+		State: storepb.HeadersLazyLoadedTrackerState{
+			UserId: "anonymous",
+		},
+	}
+	// Note that we are creating a ReaderPool that doesn't run a background cleanup task for idle
+	// Reader instances. We'll manually invoke the cleanup task when we need it as part of this test.
+	pool := newReaderPool(log.NewNopLogger(), true, idleTimeout, metrics, lazyLoadedTracker)
+	defer pool.Close()
+
+	r, err := pool.NewBinaryReader(ctx, log.NewNopLogger(), bkt, tmpDir, blockID, 3, Config{})
+	require.NoError(t, err)
+	defer func() { require.NoError(t, r.Close()) }()
+
+	// Ensure it can read data.
+	labelNames, err := r.LabelNames()
+	require.NoError(t, err)
+	require.Equal(t, []string{"a"}, labelNames)
+	require.Equal(t, float64(1), promtestutil.ToFloat64(metrics.lazyReader.loadCount))
+	require.Equal(t, float64(0), promtestutil.ToFloat64(metrics.lazyReader.unloadCount))
+
+	// lazyLoaded tracker will track the lazyLoadedBlocks into persistent file
+	pool.lazyLoadedTracker.CopyLazyLoadedState(pool.lazyReaders)
+	err = pool.lazyLoadedTracker.Persist()
+	require.NoError(t, err)
+	require.Condition(t, func() bool { return pool.lazyLoadedTracker.State.LazyLoadedBlocks[blockID.String()] > 0 }, "lazyLoadedBlocks state must be set")
+
+	// Wait enough time before checking it.
+	time.Sleep(idleTimeout * 2)
+	pool.closeIdleReaders()
+
+	// lazyLoaded tracker will remove the track from persistent file
+	pool.lazyLoadedTracker.CopyLazyLoadedState(pool.lazyReaders)
+	err = pool.lazyLoadedTracker.Persist()
+	require.NoError(t, err)
+	require.Condition(t, func() bool { return pool.lazyLoadedTracker.State.LazyLoadedBlocks[blockID.String()] == 0 }, "lazyLoadedBlocks state must be unset")
+}
+
+func prepareReaderPool(t *testing.T) (context.Context, string, *filesystem.Bucket, ulid.ULID, *ReaderPoolMetrics) {
+	ctx := context.Background()
+
+	tmpDir, err := os.MkdirTemp("", "test-indexheader")
+	require.NoError(t, err)
```

Comment:
```suggestion
	tmpDir := t.TempDir()
```

This gets cleaned up after the test and there's no need to do assertions